src/vs/workbench/api/browser/mainThreadCustomEditors.ts

```diff
@@ -106,6 +106,23 @@ export class MainThreadCustomEditors extends Disposable implements extHostProtoc
 
 		// Working copy operations
 		this._register(workingCopyFileService.onWillRunWorkingCopyFileOperation(async e => this.onWillRunWorkingCopyFileOperation(e)));
+
+		this._register(extensionService.onWillStop(e => {
+			const dirtyCustomEditors = workingCopyService.workingCopies.filter(workingCopy => {
+				return workingCopy instanceof MainThreadCustomEditorModel && workingCopy.isDirty();
+			});
+			if (!dirtyCustomEditors.length) {
+				return;
+			}
+
+			e.veto((async () => {
+				let didSaveAll = true;
+				for (const dirtyCustomEditor of dirtyCustomEditors) {
+					didSaveAll &&= await dirtyCustomEditor.save();
+				}
+				return !didSaveAll;
```

Comment:
I wonder if this is obvious to the user, should we show a dialog from here telling the user that the restart was prevented because of unsaved custom editors? Or should the extension host service show that dialog?

This is related to https://github.com/microsoft/vscode/issues/180514#issuecomment-1548176634

Reply:
Good suggestion. I will add an error notification if we veto the restart. However maybe the `veto` call in `onWillStop` should take a `reason` string and show this error itself?